morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+                        used as a axoninput
+    '''
+    df = pd.DataFrame([morph.row for morph in morphologies], columns=COLUMNS)
+    df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS}, copy=False)
+    return df
+
+
+class MorphDB(object):
+    '''A MorphInfo container.
+    It takes care of maintaining unicity of the MorphInfo element
+    and methods to write neurondb to various format (xml, dat, csv)
+    '''
+
+    def __init__(self):
+        self.df = pd.DataFrame(columns=COLUMNS)
+        self.df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS}, copy=False)
+
+    @classmethod
+    def from_neurondb(cls,
+                      neurondb: Path = None,
+                      label: str = 'default',
+                      morphology_folder: Optional[Path] = None):
+        '''Builds a MorphologyDB from a neurondb.(xml|dat) file
+        Args:
+            neurondb: path to a neurondb.(xml|dat) file
+            label: a unique label to mark all morphologies coming from this neurondb
+            morphology_folder: the location of the morphology files, if None it will default
+                to the neurondb folder
+
+        ..note:: missing keys are filled with `True` values
+        '''
+        obj = MorphDB()
+
+        if not morphology_folder:
+            morphology_folder = neurondb.parent.resolve()
+
+        if neurondb.suffix.lower() == '.dat':
+            columns = ['name', 'layer', 'mtype']
+            obj.df = pd.read_csv(neurondb, sep=r'\s+', names=columns, usecols=range(len(columns)))
+
+            fulltypes = obj.df.mtype.str.split(':', n=1, expand=True).fillna('')
+            if len(fulltypes.columns) > 1:
+                obj.df['mtype_no_subtype'] = fulltypes[0]
+                obj.df['msubtype'] = fulltypes[1]
+            else:
+                obj.df['mtype_no_subtype'] = obj.df.mtype
+                obj.df['msubtype'] = ''
+            obj.df['label'] = label
+            for missing_col in set(COLUMNS) - set(obj.df.columns):
+                obj.df[missing_col] = None
+            obj.df.layer = obj.df.layer.astype('str')
+            obj.df['path'] = obj.df.apply(lambda row: find_morph(
+                morphology_folder, row['name']), axis=1)
+            obj.df = obj.df.reindex(columns=COLUMNS)
+            for key in BOOLEAN_REPAIR_ATTRS:
+                obj.df[key] = True
+            obj.df['axon_inputs'] = [[] for _ in range(len(obj.df))]
+
+            return obj
+
+        with neurondb.open() as fd:
+            content = fd.read()
+            neurondb = xmltodict.parse(content)
+
+        morphologies = neurondb['neurondb']['listing']['morphology']
+
+        # Case where there is a single <morphology></morphology> tag in the XML
+        if not isinstance(morphologies, list):
+            morphologies = [morphologies]
+        morphologies = filter(None, morphologies)
+        morphologies = list(map(MorphInfo._from_xml, morphologies))  # noqa, pylint: disable=protected-access
```

Comment:
can't _from_xml be public?

Reply:
Nope. `from_neurondb` is public so that's what should be used.